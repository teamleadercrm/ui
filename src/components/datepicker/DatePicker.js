import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import DayPicker from 'react-day-picker';
import Box, { omitBoxProps, pickBoxProps } from '../box';
import NavigationBar from './NavigationBar';
import WeekDay from './WeekDay';
import MonthPicker from './MonthPicker';
import { convertModifiersToClassnames } from './utils';
import cx from 'classnames';
import theme from './theme.css';
import uiUtilities from '@teamleader/ui-utilities';
import LocaleUtils from './localeUtils';

/** @type {React.ComponentType<any>} */
class DatePicker extends PureComponent {
  state = {
    selectedDate: null,
    selectedMonth: null,
  };

  static getDerivedStateFromProps(props, state) {
    if (props.selectedDate !== undefined && props.selectedDate !== state.selectedDate) {
      return {
        selectedDate: props.selectedDate,
      };
    }

    return null;
  }

  handleDayClick = (day, modifiers = {}) => {
    if (modifiers[theme['disabled']]) {
      return;
    }

    this.setState(
      {
        selectedDate: day,
      },
      () => this.props.onChange(day),
    );
  };

  handleYearMonthChange = (selectedMonth) => {
    this.setState({ selectedMonth });
  };

  getMonthPickerSize = () => {
    const { showWeekNumbers, size } = this.props;
    const monthPickerSizeByDatePickerSize = {
      small: 'smallest',
      medium: showWeekNumbers ? 'medium' : 'small',
      large: 'large',
    };
    return monthPickerSizeByDatePickerSize[size];
  };

  render() {
    const { bordered, className, style, modifiers, size, withMonthPicker, showWeekNumbers, ...others } = this.props;
    const { selectedDate, selectedMonth } = this.state;
    const boxProps = pickBoxProps(others);
    const restProps = omitBoxProps(others);
    const classNames = cx(
      uiUtilities['reset-font-smoothing'],
      theme['date-picker'],
      theme[`is-${size}`],
      {
        [theme['is-bordered']]: bordered,
      },
      className,
    );

    return (
      <Box {...boxProps}>
        <DayPicker
          {...restProps}
          localeUtils={LocaleUtils}
          initialMonth={selectedDate}
          month={selectedMonth}
          className={classNames}
          style={style}
          classNames={theme}
          modifiers={convertModifiersToClassnames(modifiers, theme)}
          navbarElement={<NavigationBar size={size} withMonthPicker={withMonthPicker} />}
          onDayClick={this.handleDayClick}
          selectedDays={selectedDate}
          weekdayElement={<WeekDay size={size} />}
          showWeekNumbers={showWeekNumbers}
          captionElement={
            withMonthPicker
              ? ({ date, locale, localeUtils }) => (
                  <MonthPicker
                    date={date}
                    locale={locale}
                    localeUtils={localeUtils}
                    onChange={this.handleYearMonthChange}
                    size={this.getMonthPickerSize()}
                  />
                )
              : undefined
          }
        />
      </Box>
    );
  }
}

DatePicker.propTypes = {
  /** If true we give a border to our wrapper. */
  bordered: PropTypes.bool,
  /** A class name for the DatePicker to give custom styles. */
  className: PropTypes.string,
  /** The modifiers of the DatePicker component. */
  modifiers: PropTypes.object,
  /** Callback function that is fired when the date has changed. */
  onChange: PropTypes.func,
  /** The current selected date. */
  selectedDate: PropTypes.instanceOf(Date),
  /** Size of the DatePicker component. */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
};

DatePicker.defaultProps = {
  bordered: true,
  size: 'medium',
};

export default DatePicker;
