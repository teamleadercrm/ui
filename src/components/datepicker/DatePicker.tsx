import React, { useState, useEffect } from 'react';
import DayPicker, { DayModifiers, DayPickerProps } from 'react-day-picker';
import Box, { pickBoxProps } from '../box';
import NavigationBar from './NavigationBar';
import WeekDay from './WeekDay';
import MonthPicker from './MonthPicker';
import { convertModifiersToClassnames } from './utils';
import cx from 'classnames';
import theme from './theme.css';
import uiUtilities from '@teamleader/ui-utilities';
import { BoxProps } from '../box/Box';
import localeUtils from './localeUtils';
import { GenericComponent } from '../../@types/types';
import { SIZES } from '../../constants';

export interface DatePickerProps extends Omit<BoxProps & DayPickerProps, 'size' | 'onChange' | 'modifiers' | 'ref'> {
  /** If true we give a border to our wrapper. */
  bordered?: boolean;
  /** A class name for the DatePicker to give custom styles. */
  className?: string;
  /** The modifiers of the DatePicker component. */
  modifiers?: DayModifiers;
  /** Callback function that is fired when the date has changed. */
  onChange?: (day: Date) => void;
  /** The current selected date. */
  selectedDate?: Date;
  /** Size of the DatePicker component. */
  size?: Exclude<(typeof SIZES)[number], 'tiny' | 'fullscreen' | 'hero'>;
  /** Show a dropdown for the month? */
  withMonthPicker?: boolean;
  /** Show week numbers? */
  showWeekNumbers?: boolean;
  /** The initial month to display if no date is selected. */
  initialMonth?: Date;
}

const DatePicker: GenericComponent<DatePickerProps> = ({
  bordered = true,
  className,
  modifiers,
  size = 'medium',
  withMonthPicker,
  showWeekNumbers,
  initialMonth,
  onChange,
  ...others
}) => {
  const [selectedDate, setSelectedDate] = useState<Date | undefined>(others.selectedDate);
  const [selectedMonth, setSelectedMonth] = useState<Date>();

  useEffect(() => {
    setSelectedDate(others.selectedDate);
    setSelectedMonth(others.selectedDate);
  }, [others.selectedDate]);

  const handleDayClick = (day: Date, modifiers: DayModifiers) => {
    if (modifiers[theme['disabled']]) {
      return;
    }

    setSelectedDate(day);
    onChange && onChange(day);
  };

  const handleYearMonthChange = (selectedMonth: Date) => {
    setSelectedMonth(selectedMonth);
  };

  const getMonthPickerSize = () => {
    const monthPickerSizeByDatePickerSize: Record<
      string,
      Exclude<(typeof SIZES)[number], 'tiny' | 'fullscreen' | 'hero'>
    > = {
      small: 'smallest',
      medium: showWeekNumbers ? 'medium' : 'small',
      large: 'large',
    };

    return monthPickerSizeByDatePickerSize[size];
  };

  const classNames = cx(
    uiUtilities['reset-font-smoothing'],
    theme['date-picker'],
    theme[`is-${size}`],
    {
      [theme['is-bordered']]: bordered,
    },
    className,
  );

  return (
    <Box {...pickBoxProps(others)} data-teamleader-ui="date-picker">
      <DayPicker
        {...others}
        localeUtils={localeUtils}
        initialMonth={others.selectedDate ?? initialMonth}
        month={selectedMonth}
        className={classNames}
        classNames={theme as any}
        modifiers={modifiers && convertModifiersToClassnames(modifiers, theme)}
        navbarElement={({ ...props }) => <NavigationBar {...props} size={size} withMonthPicker={withMonthPicker} />}
        onDayClick={handleDayClick}
        selectedDays={selectedDate}
        weekdayElement={({ ...props }) => <WeekDay {...props} size={size} />}
        showWeekNumbers={showWeekNumbers}
        fixedWeeks
        captionElement={
          withMonthPicker
            ? ({ date, locale, localeUtils }) => (
                <MonthPicker
                  date={date}
                  locale={locale}
                  localeUtils={localeUtils}
                  onChange={handleYearMonthChange}
                  size={getMonthPickerSize()}
                />
              )
            : undefined
        }
      />
    </Box>
  );
};

export default DatePicker;
