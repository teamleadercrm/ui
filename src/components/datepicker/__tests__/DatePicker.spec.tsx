import React from 'react';
import { render } from '@testing-library/react';
import DatePicker from '../DatePicker';

describe('Component - DatePicker', () => {
  it('renders', async () => {
    const { asFragment } = render(<DatePicker selectedDate={new Date('1965-03-15')} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
