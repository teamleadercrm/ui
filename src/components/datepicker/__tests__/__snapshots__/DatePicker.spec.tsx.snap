// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - DatePicker renders 1`] = `
<DocumentFragment>
  <div
    class="box"
    data-teamleader-ui="date-picker"
  >
    <div
      class="container reset-font-smoothing date-picker is-medium is-bordered"
      lang="en"
    >
      <div
        class="wrapper"
        tabindex="0"
      >
        <div
          class="box display-flex justify-content-space-between navBar"
        >
          <button
            class="box button-base icon-button is-small"
            data-teamleader-ui="icon-button"
            title="February"
            type="button"
          >
            <span
              class="box align-items-center display-flex teal darkest"
              data-teamleader-ui="icon"
            >
              <svg
                fill="currentColor"
                height="14"
                preserveAspectRatio="xMidYMid meet"
                style="opacity: 0.84; vertical-align: middle;"
                viewBox="0 0 14 14"
                width="14"
              >
                <path
                  clip-rule="evenodd"
                  d="M4.575 5.955l1.158-1.171a1.054 1.054 0 000-1.478 1.024 1.024 0 00-1.46 0L.623 7l3.65 3.694a1.024 1.024 0 001.46 0 1.054 1.054 0 000-1.478L4.575 8.045h7.015c.57 0 1.033-.468 1.033-1.045s-.463-1.045-1.033-1.045H4.575z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
          <button
            class="box button-base icon-button is-small"
            data-teamleader-ui="icon-button"
            title="April"
            type="button"
          >
            <span
              class="box align-items-center display-flex teal darkest"
              data-teamleader-ui="icon"
            >
              <svg
                fill="currentColor"
                height="14"
                preserveAspectRatio="xMidYMid meet"
                style="opacity: 0.84; vertical-align: middle;"
                viewBox="0 0 14 14"
                width="14"
              >
                <path
                  clip-rule="evenodd"
                  d="M9.047 5.955L7.89 4.784a1.054 1.054 0 010-1.478 1.024 1.024 0 011.46 0L13 7l-3.65 3.694a1.024 1.024 0 01-1.46 0 1.054 1.054 0 010-1.478l1.157-1.171H2.032C1.462 8.045 1 7.577 1 7s.462-1.045 1.032-1.045h7.015z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="months"
        >
          <div
            class="month"
            role="grid"
          >
            <div
              aria-live="polite"
              class="caption"
              role="heading"
            >
              <div>
                March 1965
              </div>
            </div>
            <div
              class="weekdays"
              role="rowgroup"
            >
              <div
                class="weekdaysRow"
                role="row"
              >
                <div
                  class="weekday"
                  title="Monday"
                >
                  M
                </div>
                <div
                  class="weekday"
                  title="Tuesday"
                >
                  T
                </div>
                <div
                  class="weekday"
                  title="Wednesday"
                >
                  W
                </div>
                <div
                  class="weekday"
                  title="Thursday"
                >
                  T
                </div>
                <div
                  class="weekday"
                  title="Friday"
                >
                  F
                </div>
                <div
                  class="weekday"
                  title="Saturday"
                >
                  S
                </div>
                <div
                  class="weekday"
                  title="Sunday"
                >
                  S
                </div>
              </div>
            </div>
            <div
              class="body"
              role="rowgroup"
            >
              <div
                class="week"
                role="row"
              >
                <div
                  aria-disabled="false"
                  aria-label="Monday, March 1, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="0"
                >
                  1
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Tuesday, March 2, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  2
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Wednesday, March 3, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  3
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Thursday, March 4, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  4
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Friday, March 5, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  5
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Saturday, March 6, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  6
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Sunday, March 7, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  7
                </div>
              </div>
              <div
                class="week"
                role="row"
              >
                <div
                  aria-disabled="false"
                  aria-label="Monday, March 8, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  8
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Tuesday, March 9, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  9
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Wednesday, March 10, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  10
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Thursday, March 11, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  11
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Friday, March 12, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  12
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Saturday, March 13, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  13
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Sunday, March 14, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  14
                </div>
              </div>
              <div
                class="week"
                role="row"
              >
                <div
                  aria-disabled="false"
                  aria-label="Monday, March 15, 1965"
                  aria-selected="true"
                  class="day selected"
                  role="gridcell"
                  tabindex="-1"
                >
                  15
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Tuesday, March 16, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  16
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Wednesday, March 17, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  17
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Thursday, March 18, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  18
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Friday, March 19, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  19
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Saturday, March 20, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  20
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Sunday, March 21, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  21
                </div>
              </div>
              <div
                class="week"
                role="row"
              >
                <div
                  aria-disabled="false"
                  aria-label="Monday, March 22, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  22
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Tuesday, March 23, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  23
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Wednesday, March 24, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  24
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Thursday, March 25, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  25
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Friday, March 26, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  26
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Saturday, March 27, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  27
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Sunday, March 28, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  28
                </div>
              </div>
              <div
                class="week"
                role="row"
              >
                <div
                  aria-disabled="false"
                  aria-label="Monday, March 29, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  29
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Tuesday, March 30, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  30
                </div>
                <div
                  aria-disabled="false"
                  aria-label="Wednesday, March 31, 1965"
                  aria-selected="false"
                  class="day "
                  role="gridcell"
                  tabindex="-1"
                >
                  31
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Thursday, April 1, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  1
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Friday, April 2, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  2
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Saturday, April 3, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  3
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Sunday, April 4, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  4
                </div>
              </div>
              <div
                class="week"
                role="row"
              >
                <div
                  aria-disabled="true"
                  aria-label="Monday, April 5, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  5
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Tuesday, April 6, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  6
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Wednesday, April 7, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  7
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Thursday, April 8, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  8
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Friday, April 9, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  9
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Saturday, April 10, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  10
                </div>
                <div
                  aria-disabled="true"
                  aria-label="Sunday, April 11, 1965"
                  aria-selected="false"
                  class="day outside"
                  role="gridcell"
                  tabindex="-1"
                >
                  11
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
